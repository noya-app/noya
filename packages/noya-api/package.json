{
  "name": "noya-api",
  "private": false,
  "modular": {
    "type": "package"
  },
  "main": "./src/index.ts",
  "version": "1.0.0",
  "dependencies": {
    "@legendapp/state": "^2.1.1"
  }
}
