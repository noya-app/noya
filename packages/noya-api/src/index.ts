import { NoyaClient, NoyaImageGenerator } from './core/client';
import { NoyaAPIError } from './core/error';
import { NoyaLocalStorageClient } from './core/localStorageClient';
import { NoyaMemoryClient } from './core/memoryClient';
import { NoyaNetworkClient } from './core/networkClient';
import {
  DS,
  DSConfig,
  DSSource,
  NoyaExportFormat,
  NoyaFile,
  NoyaFileData,
  NoyaFileList,
  NoyaGeneratedName,
  NoyaJson,
  NoyaPrice,
  NoyaProduct,
  NoyaRandomIconResponse,
  NoyaRandomImageResponse,
  NoyaSession,
  NoyaShare,
  NoyaSharedFile,
  NoyaSubscription,
  NoyaSubscriptionItem,
  NoyaUser,
} from './core/schema';

export * from './react/context';
export * from './react/hooks';

export namespace NoyaAPI {
  export const NetworkClient = NoyaNetworkClient;
  export type NetworkClient = NoyaNetworkClient;

  export const Client = NoyaClient;
  export type Client = NoyaClient;

  export const Error = NoyaAPIError;
  export type Error = NoyaAPIError;

  export const MemoryClient = NoyaMemoryClient;
  export type MemoryClient = NoyaMemoryClient;

  export const LocalStorageClient = NoyaLocalStorageClient;
  export type LocalStorageClient = NoyaLocalStorageClient;

  export type Json = NoyaJson;
  export type File = NoyaFile;
  export type FileData = NoyaFileData;
  export type FileList = NoyaFileList;
  export type User = NoyaUser;
  export type Session = NoyaSession;
  export type ExportFormat = NoyaExportFormat;
  export type Subscription = NoyaSubscription;
  export type Product = NoyaProduct;
  export type Price = NoyaPrice;
  export type SubscriptionItem = NoyaSubscriptionItem;
  export type Share = NoyaShare;
  export type SharedFile = NoyaSharedFile;
  export type GeneratedName = NoyaGeneratedName;
  export type RandomImageResponse = NoyaRandomImageResponse;
  export type RandomIconResponse = NoyaRandomIconResponse;
  export type ImageGenerator = NoyaImageGenerator;
}

export type { DS, DSConfig, DSSource };
