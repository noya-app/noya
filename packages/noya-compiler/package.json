{
  "name": "noya-compiler",
  "private": false,
  "modular": {
    "type": "package"
  },
  "main": "./src/index.ts",
  "version": "1.0.0",
  "dependencies": {
    "codesandbox": "^2.2.3"
  },
  "optionalDependencies": {
    "@dabbott/patterns": "^0.0.4"
  }
}
