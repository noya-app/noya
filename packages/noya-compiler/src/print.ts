import prettier from 'prettier';
import prettierTypeScript from 'prettier/parser-typescript';
import ts from 'typescript';

export function print(nodes: ts.Node | ts.Node[]) {
  const sourceFile = ts.createSourceFile(
    'App.tsx',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TSX,
  );

  const printer = ts.createPrinter();

  const source = Array.isArray(nodes)
    ? printer.printList(
        ts.ListFormat.MultiLine,
        ts.factory.createNodeArray(nodes),
        sourceFile,
      )
    : printer.printNode(ts.EmitHint.Unspecified, nodes, sourceFile);

  return source;
}

export function format(text: string) {
  return prettier.format(text, {
    singleQuote: true,
    trailingComma: 'es5',
    printWidth: 80,
    proseWrap: 'always',
    parser: 'typescript',
    plugins: [prettierTypeScript],
  });
}
