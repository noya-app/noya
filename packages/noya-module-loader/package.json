{
  "name": "noya-module-loader",
  "private": false,
  "modular": {
    "type": "package"
  },
  "main": "./src/index.ts",
  "version": "1.0.0"
}
