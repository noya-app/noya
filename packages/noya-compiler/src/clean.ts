import ts from 'typescript';
import { removeEmptyClassNames } from './passes/removeEmptyClassNames';
import { removeEmptyStyles } from './passes/removeEmptyStyles';
import { removeUndefinedStyles } from './passes/removeUndefinedStyles';
import { format, print } from './print';

export function clean(text: string) {
  const sourceFile = ts.createSourceFile(
    'temp.tsx',
    text,
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TSX,
  );

  const updated = removeEmptyClassNames(
    removeEmptyStyles(removeUndefinedStyles(sourceFile)),
  );

  return format(print(updated));
}
