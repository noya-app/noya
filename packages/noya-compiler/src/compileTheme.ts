import {
  DesignSystemDefinition,
  Transformer,
} from '@noya-design-system/protocol';
import { generateImportDeclarations } from 'noya-compiler';
import ts from 'typescript';
import { buildNamespaceMap, createExpressionCode } from './common';
import { format, print } from './print';

/**
 * Takes an input like:
 *
 * ```ts
 * const themeTransformer = x.f(extendTheme, [
 *   x.a("theme"),
 *   x.f(withDefaultColorScheme, [{ colorScheme: "primary" }]),
 *   { config: { initialColorMode: x.a("theme.colorMode") } },
 * ])
 * ```
 *
 * And returns:
 *
 * ```ts
 * import { extendTheme } from "@chakra-ui/react";
 *
 * export const theme = extendTheme({
 *   { ...theme input object... },
 *   withDefaultColorScheme({ colorScheme: 'primary' }),
 *   { config: { initialColorMode: 'light' } }
 * });
 * ```
 */
export function generateThemeFile(
  DesignSystem: DesignSystemDefinition,
  themeValue: any,
): string {
  if (!DesignSystem.themeTransformer) return '';

  const namespaceMap = buildNamespaceMap(DesignSystem.imports);
  const imports: { name: string; source: string }[] = [];

  function convert(transformer: any): ts.Expression {
    if (isTransformer(transformer)) {
      switch (transformer.__transformer) {
        case 'access': {
          return createExpressionCode(
            accessPath(
              transformer.value ?? themeValue,
              transformer.path as string,
            ),
          );
        }
        case 'function': {
          const func = namespaceMap.get(transformer.value);

          if (!func) return ts.factory.createNull();

          imports.push(func);

          return ts.factory.createCallExpression(
            ts.factory.createIdentifier(func.name),
            undefined,
            transformer.args.map(convert),
          );
        }
      }
    }

    if (Array.isArray(transformer)) {
      return ts.factory.createArrayLiteralExpression(
        transformer.map((item) => convert(item)),
      );
    }

    if (typeof transformer === 'object' && transformer !== null) {
      return ts.factory.createObjectLiteralExpression(
        Object.entries(transformer).flatMap(([key, value]) => {
          const expression = convert(value);

          return [
            ts.factory.createPropertyAssignment(
              ts.factory.createIdentifier(key),
              expression,
            ),
          ];
        }),
      );
    }

    return createExpressionCode(transformer);
  }

  const ast = convert(DesignSystem.themeTransformer);

  const themeExportAst = ts.factory.createVariableStatement(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          ts.factory.createIdentifier('theme'),
          undefined,
          undefined,
          ast,
        ),
      ],
      ts.NodeFlags.Const,
    ),
  );

  const importsAst = generateImportDeclarations(imports);

  return format([print(importsAst), print(themeExportAst)].join('\n\n'));
}

function accessPath(data: any, path: string): any {
  if (path === '.') return data;
  const parts = path.split('.');
  let currentValue = data;
  for (let part of parts) {
    currentValue = currentValue[part];
  }
  return currentValue;
}

function isTransformer(value: unknown): value is Transformer {
  return (
    typeof value === 'object' &&
    value !== null &&
    '__transformer' in value &&
    (value.__transformer === 'function' || value.__transformer === 'access')
  );
}
