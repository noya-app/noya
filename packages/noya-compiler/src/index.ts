import {
  component,
  DesignSystemDefinition,
} from '@noya-design-system/protocol';
import Sketch from 'noya-file-format';
import { loadDesignSystem } from 'noya-module-loader';
import { BlockDefinition, Layers } from 'noya-state';
import { unique } from 'noya-utils';
import prettier from 'prettier';
import prettierTypeScript from 'prettier/parser-typescript';
import React, { isValidElement } from 'react';
import { flat } from 'tree-visit';
import ts from 'typescript';
import { removeEmptyStyles } from './removeEmptyStyles';
import { removeUndefinedStyles } from './removeUndefinedStyles';
import { escapeHtml, sanitizePackageName } from './validate';

function createExpressionCode(value: unknown): ts.Expression {
  if (isSimpleElement(value)) {
    return createElementCode(value);
  }

  switch (typeof value) {
    case 'string':
      return ts.factory.createStringLiteral(value);
    case 'number':
      return ts.factory.createNumericLiteral(value);
    case 'boolean':
      return value ? ts.factory.createTrue() : ts.factory.createFalse();
    case 'object':
      if (value === null) return ts.factory.createNull();

      if (Array.isArray(value)) {
        return ts.factory.createArrayLiteralExpression(
          value.map((item) => createExpressionCode(item)),
        );
      }

      return ts.factory.createObjectLiteralExpression(
        Object.entries(value).flatMap(([key, value]) => {
          const expression = createExpressionCode(value);

          return [
            ts.factory.createPropertyAssignment(
              ts.factory.createIdentifier(key),
              expression,
            ),
          ];
        }),
      );
    case 'undefined':
      return ts.factory.createIdentifier('undefined');
    default:
      return ts.factory.createNull();
  }
}

function createJsxElement(
  openingElement: ts.JsxOpeningElement,
  children: readonly ts.JsxChild[],
) {
  if (children.length === 0) {
    return ts.factory.createJsxSelfClosingElement(
      openingElement.tagName,
      openingElement.typeArguments,
      openingElement.attributes,
    );
  }

  return ts.factory.createJsxElement(
    openingElement,
    children,
    ts.factory.createJsxClosingElement(openingElement.tagName),
  );
}

const simpleElementSymbol = Symbol('simpleElement');

function isSimpleElement(value: unknown): value is SimpleElement {
  return (
    typeof value === 'object' && value !== null && simpleElementSymbol in value
  );
}

type SimpleElement = {
  [simpleElementSymbol]: true;
  name: string;
  source?: string;
  props: Record<string, unknown>;
  children: (SimpleElement | string)[];
};

export function createElementCode({
  name,
  props,
  children,
}: SimpleElement): ts.JsxElement | ts.JsxSelfClosingElement {
  return createJsxElement(
    ts.factory.createJsxOpeningElement(
      ts.factory.createIdentifier(name),
      undefined,
      ts.factory.createJsxAttributes(
        Object.entries(props).flatMap(([key, value]) => {
          const expression = createExpressionCode(value);

          if (!expression) return [];

          return [
            ts.factory.createJsxAttribute(
              ts.factory.createIdentifier(key),
              expression.kind === ts.SyntaxKind.TrueKeyword
                ? undefined
                : expression.kind === ts.SyntaxKind.StringLiteral
                ? (expression as ts.StringLiteral)
                : ts.factory.createJsxExpression(undefined, expression),
            ),
          ];
        }),
      ),
    ),
    children.map((child) =>
      typeof child === 'string'
        ? isSafeForJsxText(child)
          ? ts.factory.createJsxText(child, false)
          : ts.factory.createJsxExpression(
              undefined,
              ts.factory.createStringLiteral(child),
            )
        : createElementCode(child),
    ),
  );
}

function isSafeForJsxText(text: string) {
  return !/[{}<>]/.test(text);
}

export interface CompilerConfiguration {
  name: string;
  artboard: Sketch.Artboard;
  Blocks: Record<string, BlockDefinition>;
  DesignSystem: string | DesignSystemDefinition;
  target: 'standalone' | 'codesandbox';
}

function findElementNameAndSource(
  element: React.ReactElement,
  DesignSystem: DesignSystemDefinition,
  Components: Map<React.ComponentType, { name: string; source?: string }>,
):
  | {
      name: string;
      element: React.ReactElement;
      source?: string;
    }
  | undefined {
  // This is a DOM element
  if (typeof element.type === 'string') {
    return { name: element.type, element };
  }

  // This is a component exported directly from the design system
  const component = Components.get(element.type);

  if (component) {
    return { ...component, element };
  }

  const protocolComponent = Object.values(DesignSystem.components).find(
    (value) => value === element.type,
  );

  // This is an adapter function that returns a DOM or design system component
  const libraryElement = protocolComponent?.(element.props);

  if (!isValidElement(libraryElement)) return;

  return findElementNameAndSource(libraryElement, DesignSystem, Components);
}

function createSimpleElement(
  originalElement: React.ReactElement,
  DesignSystem: DesignSystemDefinition,
): SimpleElement | undefined {
  const Components = buildComponentMap(DesignSystem.imports);

  const elementType = findElementNameAndSource(
    originalElement,
    DesignSystem,
    Components,
  );

  if (!elementType) return;

  const { element, name, source } = elementType;

  return {
    [simpleElementSymbol]: true,
    name,
    source,
    // Filter out children prop and undefined props
    props: Object.fromEntries(
      Object.entries(element.props)
        .filter(
          ([key, value]) =>
            key !== 'children' &&
            !key.startsWith('data-noya-') &&
            value !== undefined,
        )
        .map(([key, value]) => {
          if (React.isValidElement(value)) {
            return [key, createSimpleElement(value, DesignSystem)];
          }
          return [key, value];
        }),
    ),
    children: React.Children.toArray(element.props.children).flatMap(
      (element): (SimpleElement | string)[] => {
        if (typeof element === 'string' && element !== '') return [element];
        const validElement = React.isValidElement(element);
        if (!validElement) return [];
        const mapped = createSimpleElement(element, DesignSystem);
        return mapped ? [mapped] : [];
      },
    ),
  };
}

export function mapBlockToElement(
  {
    Blocks,
    DesignSystem,
  }: {
    Blocks: CompilerConfiguration['Blocks'];
    DesignSystem: DesignSystemDefinition;
  },
  layer: Sketch.SymbolInstance,
): SimpleElement | undefined {
  const block = Blocks[layer.symbolID];

  if (!block) return;

  const element = block.render(
    {
      h: DesignSystem.createElement as any,
      Components: DesignSystem.components,
    },
    {
      layer: layer,
      frame: layer.frame,
      symbolId: layer.symbolID,
      blockText: layer.blockText,
      resolvedBlockData: layer.resolvedBlockData,
      getBlock: (id) => Blocks[id],
      overrideValues: layer.overrideValues,
      dataSet: {
        id: layer.do_objectID,
        parentId: layer.do_objectID,
      },
    },
  );

  if (!element || !isValidElement(element)) return;

  const root = createSimpleElement(element, DesignSystem);

  if (!root) return;

  return {
    [simpleElementSymbol]: true,
    name: 'Frame',
    props: {
      ...(layer.frame.x !== 0 && { left: layer.frame.x }),
      ...(layer.frame.y !== 0 && { top: layer.frame.y }),
      width: layer.frame.width,
      height: layer.frame.height,
    },
    children: [root],
  };
}

function buildComponentMap(imports: DesignSystemDefinition['imports']) {
  const Components = new Map<any, { name: string; source: string }>();

  for (const declaration of imports ?? []) {
    for (let [name, value] of Object.entries(declaration.namespace)) {
      Components.set(value, { name, source: declaration.source });
    }
  }

  return Components;
}

// We should look this up more dynamically the same way we do for blocks
function findSourceByName(
  DesignSystem: DesignSystemDefinition,
  name: string,
): { source: string; name: string } | undefined {
  const importDeclaration = DesignSystem.imports?.find(({ namespace }) =>
    Object.keys(namespace).includes(name),
  );

  if (!importDeclaration) return;

  return {
    source: importDeclaration.source,
    name,
  };
}

export async function compile(configuration: CompilerConfiguration) {
  const { artboard } = configuration;

  const DesignSystem =
    typeof configuration.DesignSystem === 'string'
      ? await loadDesignSystem(configuration.DesignSystem)
      : configuration.DesignSystem;

  const components = artboard.layers
    .filter(Layers.isSymbolInstance)
    .flatMap((layer) => {
      const element = mapBlockToElement(
        {
          Blocks: configuration.Blocks,
          DesignSystem,
        },
        layer,
      );

      return element ? [element] : [];
    });

  const getChildren = (
    element: SimpleElement | string,
  ): (string | SimpleElement)[] => {
    if (typeof element === 'string') return [];

    return [
      ...element.children,
      ...Object.values(element.props).filter(isSimpleElement),
    ];
  };

  const ProviderComponent = DesignSystem.components[component.id.Provider];

  const providerElement = createSimpleElement(
    DesignSystem.createElement(ProviderComponent),
    DesignSystem,
  );

  if (providerElement) {
    providerElement.children = components;
  }

  const fakeRoot: SimpleElement = providerElement ?? {
    [simpleElementSymbol]: true,
    name: 'Frame',
    children: components,
    props: {},
  };

  const boxSource = findSourceByName(DesignSystem, 'Box');

  const imports = (DesignSystem.imports ?? []).flatMap(
    ({ source, alwaysInclude }) => {
      const names = unique([
        ...flat(fakeRoot, { getChildren }).flatMap((element) =>
          typeof element !== 'string' && element.source === source
            ? [element.name]
            : [],
        ),
        ...(boxSource?.source === source ? [boxSource.name] : []),
      ]);

      if (names.length === 0 && !alwaysInclude) return [];

      return [
        ts.factory.createImportDeclaration(
          undefined,
          names.length === 0
            ? undefined
            : ts.factory.createImportClause(
                false,
                undefined,
                ts.factory.createNamedImports(
                  names.map((name) =>
                    ts.factory.createImportSpecifier(
                      false,
                      undefined,
                      ts.factory.createIdentifier(name),
                    ),
                  ),
                ),
              ),
          ts.factory.createStringLiteral(source),
        ),
      ];
    },
  );

  const func = ts.factory.createFunctionDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    undefined,
    'App',
    undefined,
    [],
    undefined,
    ts.factory.createBlock([
      ts.factory.createReturnStatement(
        providerElement
          ? createElementCode(providerElement)
          : ts.factory.createJsxFragment(
              ts.factory.createJsxOpeningFragment(),
              components.map(createElementCode),
              ts.factory.createJsxJsxClosingFragment(),
            ),
      ),
    ]),
  );

  const allDependencies = (DesignSystem.imports ?? []).reduce(
    (result, importDeclaration) => ({
      ...result,
      ...importDeclaration.dependencies,
    }),
    DesignSystem.dependencies ?? {},
  );

  const allDevDependencies = (DesignSystem.imports ?? []).reduce(
    (result, importDeclaration) => ({
      ...result,
      ...importDeclaration.devDependencies,
    }),
    DesignSystem.devDependencies ?? {},
  );

  const frameComponent = `/**
 * To make your layout responsive, delete this Frame component and replace any
 * instance of it with your own layout components that use e.g. flexbox.
 */
function Frame(
  props: React.PropsWithChildren<
    Pick<React.CSSProperties, "left" | "top" | "width" | "height">
  >
) {
  return (
    <${boxSource?.name} 
      style={{
        position: 'absolute',
        left: props.left,
        top: props.top,
        width: props.width,
        height: props.height,
      }}
    >
      {props.children}
    </${boxSource?.name}>
  )
}`;

  const files = {
    'App.tsx': clean(
      [
        print(imports),
        `import * as React from "react";`,
        frameComponent,
        print(func),
      ].join('\n\n'),
    ),
    '.postcssrc': `{
  "plugins": {
    "tailwindcss": {}
  }
}`,
    'tailwind.config.js': `module.exports = {
  content: ["./*.{html,js,jsx,ts,tsx}"],
  theme: {
    extend: {},
  },
  variants: {},
  plugins: [],
};`,
    'index.css': `@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
`,
    'index.html': `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>${escapeHtml(configuration.name)}</title>
  </head>
  <body>
    <div id="root"></div>
    <link href="./index.css" rel="stylesheet" />
    ${
      configuration.target === 'codesandbox'
        ? '<script src="https://cdn.tailwindcss.com"></script>'
        : ''
    }
    <script type="module" src="index.tsx"></script>
  </body>
</html>`,
    'index.tsx': `import { createRoot } from "react-dom/client";
import * as React from 'react';

import { App } from "./App";

const rootElement = document.getElementById("root");
const root = createRoot(rootElement!);

root.render(<App />);`,
    'package.json': JSON.stringify(
      {
        name: sanitizePackageName(configuration.name),
        version: '0.0.1',
        scripts: {
          dev: 'parcel index.html --open',
          build: 'parcel build index.html',
        },
        dependencies: allDependencies,
        devDependencies: {
          autoprefixer: '10.4.14',
          parcel: '^2.8.3',
          postcss: '8.4.21',
          tailwindcss: '3.2.7',
          process: '^0.11.10',
          ...allDevDependencies,
        },
      },
      null,
      2,
    ),
    'README.md': `# ${configuration.name}

## Local Development

\`\`\`
npm install
npm run dev
\`\`\`

## Production Build

\`\`\`
npm run build
\`\`\`
`,
  };

  return files;
}

export function print(nodes: ts.Node | ts.Node[]) {
  const sourceFile = ts.createSourceFile(
    'App.tsx',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TSX,
  );

  const printer = ts.createPrinter();

  const source = Array.isArray(nodes)
    ? printer.printList(
        ts.ListFormat.MultiLine,
        ts.factory.createNodeArray(nodes),
        sourceFile,
      )
    : printer.printNode(ts.EmitHint.Unspecified, nodes, sourceFile);

  return source;
}

export function format(text: string) {
  return prettier.format(text, {
    singleQuote: true,
    trailingComma: 'es5',
    printWidth: 80,
    proseWrap: 'always',
    parser: 'typescript',
    plugins: [prettierTypeScript],
  });
}

export function clean(text: string) {
  const sourceFile = ts.createSourceFile(
    'temp.tsx',
    text,
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TSX,
  );

  const updated = removeEmptyStyles(removeUndefinedStyles(sourceFile));

  return format(print(updated));
}

export * from './codesandbox';
