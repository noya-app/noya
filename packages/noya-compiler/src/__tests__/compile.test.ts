/* eslint-disable testing-library/render-result-naming-convention */
import {
  component,
  DesignSystemDefinition,
  Theme,
} from '@noya-design-system/protocol';
import { DS } from 'noya-api';
import { createElementCode, createSimpleElement } from 'noya-compiler';
import {
  createResolvedNode,
  FindComponent,
  Model,
  renderResolvedNode,
} from 'noya-component';
import { loadDesignSystem } from 'noya-module-loader';
import { tailwindColors } from 'noya-tailwind';
import { uuid } from 'noya-utils';
import { clean } from '../clean';
import { generateThemeFile } from '../compileTheme';
import { print } from '../print';

const HeroComponent = Model.component({
  name: 'Hero',
  componentID: uuid(),
  rootElement: Model.primitiveElement({
    id: 'box',
    name: 'Content',
    classNames: Model.classNames([
      'flex',
      'flex-col',
      'items-center',
      'justify-center',
      'p-10',
      'flex-1',
      'gap-4',
    ]),
    componentID: component.id.Box,
    children: [
      Model.primitiveElement({
        componentID: component.id.Tag,
        children: [Model.string('New')],
      }),
      Model.primitiveElement({
        name: 'Title',
        componentID: component.id.Text,
        children: [Model.string('Create, iterate, inspire.')],
        classNames: Model.classNames([
          'variant-h1',
          'leading-none',
          'text-center',
        ]),
      }),
      Model.primitiveElement({
        name: 'Subtitle',
        componentID: component.id.Text,
        children: [Model.string('Turn great ideas into new possibilities.')],
        classNames: Model.classNames([
          'variant-h4',
          'leading-none',
          'text-center',
        ]),
      }),
      Model.primitiveElement({
        id: 'a',
        name: 'Actions Row',
        classNames: Model.classNames(['flex', 'items-center', 'gap-4', 'mt-2']),
        componentID: component.id.Box,
        children: [
          Model.primitiveElement({
            componentID: component.id.Button,
            children: [Model.string('Get Started')],
          }),
          Model.primitiveElement({
            componentID: component.id.Link,
            children: [Model.string('Learn More')],
          }),
        ],
      }),
    ],
  }),
});

jest.setTimeout(20000);

let ChakraDesignSystem: DesignSystemDefinition;
// let MaterialDesignSystem: DesignSystemDefinition;

beforeAll(async () => {
  ChakraDesignSystem = await loadDesignSystem('chakra');
  // MaterialDesignSystem = await loadDesignSystem('mui');
});

const ds: DS = {
  components: [HeroComponent],
  config: {
    colorMode: 'light',
    colors: {
      primary: 'violet',
    },
  },
  source: {
    name: 'chakra',
    type: 'npm',
    version: '0.0.1',
  },
};

describe('renders', () => {
  const findComponent: FindComponent = (componentID) => {
    return ds.components?.find(
      (component) => component.componentID === componentID,
    );
  };

  const noyaComponent = findComponent(HeroComponent.componentID);

  if (!noyaComponent) {
    throw new Error(
      `Could not find component with id ${HeroComponent.componentID}`,
    );
  }

  const resolvedNode = createResolvedNode(
    findComponent,
    noyaComponent.rootElement,
  );

  test('default', () => {
    const reactNode = renderResolvedNode({
      contentEditable: false,
      disableTabNavigation: false,
      includeDataProps: false,
      system: ChakraDesignSystem,
      dsConfig: ds.config,
      resolvedNode,
    });

    const code = createElementCode(
      createSimpleElement(reactNode, ChakraDesignSystem)!,
    );

    const out = clean(print(code));

    expect(out).toMatchSnapshot();
  });

  test('classnames', () => {
    const reactNode = renderResolvedNode({
      contentEditable: false,
      disableTabNavigation: false,
      includeDataProps: false,
      system: ChakraDesignSystem,
      dsConfig: ds.config,
      resolvedNode,
      stylingMode: 'tailwind',
    });

    const code = createElementCode(
      createSimpleElement(reactNode, ChakraDesignSystem)!,
    );

    const out = clean(print(code));

    expect(out).toMatchSnapshot();
  });
});

describe('theme', () => {
  it('generates transformer', () => {
    const config = ds.config;

    const theme: Theme = {
      colorMode: config.colorMode ?? 'light',
      colors: {
        primary: (tailwindColors as any)[config.colors.primary],
        neutral: tailwindColors.slate,
      },
    };

    const transformer = generateThemeFile(ChakraDesignSystem, { theme });

    expect(transformer).toMatchSnapshot();
  });
});
