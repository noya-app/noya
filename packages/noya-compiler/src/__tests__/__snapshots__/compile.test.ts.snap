// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default 1`] = `
"<Box
  display=\\"flex\\"
  flexDirection=\\"column\\"
  alignItems=\\"center\\"
  justifyContent=\\"center\\"
  p=\\"2.5rem\\"
  flex=\\"1\\"
  gap=\\"1rem\\"
>
  <Tag>New</Tag>
  <Text fontSize=\\"5xl\\" lineHeight=\\"1\\" textAlign=\\"center\\">
    Create, iterate, inspire.
  </Text>
  <Text fontSize=\\"2xl\\" lineHeight=\\"1\\" textAlign=\\"center\\">
    Turn great ideas into new possibilities.
  </Text>
  <Box display=\\"flex\\" alignItems=\\"center\\" gap=\\"1rem\\" mt=\\"0.5rem\\">
    <Button>Get Started</Button>
    <Link>Learn More</Link>
  </Box>
</Box>;
"
`;

exports[`theme generates transformer 1`] = `
"import { extendTheme, withDefaultColorScheme } from '@chakra-ui/react';

export const theme = extendTheme(
  {
    colorMode: 'light',
    colors: {
      primary: {
        50: '#f5f3ff',
        100: '#ede9fe',
        200: '#ddd6fe',
        300: '#c4b5fd',
        400: '#a78bfa',
        500: '#8b5cf6',
        600: '#7c3aed',
        700: '#6d28d9',
        800: '#5b21b6',
        900: '#4c1d95',
        950: '#2e1065',
      },
      neutral: {
        50: '#f8fafc',
        100: '#f1f5f9',
        200: '#e2e8f0',
        300: '#cbd5e1',
        400: '#94a3b8',
        500: '#64748b',
        600: '#475569',
        700: '#334155',
        800: '#1e293b',
        900: '#0f172a',
        950: '#020617',
      },
    },
  },
  withDefaultColorScheme({ colorScheme: 'primary' }),
  { config: { initialColorMode: 'light' } }
);
"
`;
